nixos/modules/system/activation/top-level.nix

```diff
@@ -14,12 +14,15 @@ let
     ${
       if config.boot.initrd.enable && config.boot.initrd.systemd.enable then
         ''
-          cp ${config.system.build.bootStage2} $out/prepare-root
-          substituteInPlace $out/prepare-root --subst-var-by systemConfig $out
           # This must not be a symlink or the abs_path of the grub builder for the tests
           # will resolve the symlink and we end up with a path that doesn't point to a
           # system closure.
           cp "$systemd/lib/systemd/systemd" $out/init
+
+          ${lib.optionalString (!config.system.nixos-init.enable) ''
+            cp ${config.system.build.bootStage2} $out/prepare-root
+            substituteInPlace $out/prepare-root --subst-var-by systemConfig $out
+          ''}
```

Comment:
Is `cp "$systemd/lib/systemd/systemd" $out/init` still valid with nixos-init?

Reply:
Yes, we still systemd as init even when we use nixos-init. nixos-init only replaces prepare-root (i.e. stage-2-init.sh).